mono/utils/mono-error-internals.h

```diff
@@ -43,10 +43,34 @@ struct _MonoErrorBoxed {
 	MonoImage *image;
 };
 
-#define error_init(error) do {	\
-	((MonoErrorInternal*)(error))->error_code = MONO_ERROR_NONE;	\
-	((MonoErrorInternal*)(error))->flags = 0;	\
-} while (0);
+// FIXME This is duplicated.
+#define ERROR_DECL \
+	MonoError error_value; \
+	MonoError * const error = &error_value; \
+	error_init_internal (error);
+
+#define ERROR_DECL_NAME(x) \
+	MonoError x; \
+	error_init_internal (&x);
+
+// MonoError.hidden_0 is MonoErrorInternal.flags.
+// error_init is only now for compatibility at API boundaries.
+// Internal users should use ERROR_DECL.
+
+#undef error_init
+#define error_init(error) ((error)->error_code = (error)->hidden_0 = 0)
```

Comment:
Why not cast to `MonoErrorInternal*` and assign to `->flags`?

Reply:
Casts are to be avoided, and this one is not necessary. When you have the cast, the macro can accept anything and treat it as an error, including completely unrelated things, and corrupt memory.

Reply:
what do you think of

```c
static inline void
error_init (MonoError *error)
{
  ((MonoErrorInternals*)error)->error_code = MONO_ERROR_NONE;
  ((MonoErrorInternals*)error)->flags = 0;
}
```

Reply:
That is also good, but it seems people really like the reallyforceinline of a macro, that at one time, error_init was a function and showed up in profiling. But yes.